### PR TITLE
MueLu: refactor phase2 aggregation

### PR DESCRIPTION
**Pull request by lucbv**
@trilinos/muelu 

## Description
Adding refactor of phase 2a and 2b of aggregation using the distance 2 coloring.

## Motivation and Context
This work is part of issue #5838 and aims at having the setup of MueLu SA-AMG performed on GPU.

## Related Issues

* Closes 
* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to 
* Part of #5838 
* Composed of 

## How Has This Been Tested?
The unit-tests for kokkos are testing the new algorithms and are passing correctly

## Checklist

- [x] My commit messages mention the appropriate GitHub issue numbers.
- [x] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

**Comment by lucbv** [comment]
This is still a work in progress, the work in this branch has been tested on white but I want to see if the auto-tester has any complaints about it?
Also @brian-kelley here is the PR to which you should be able to push I think, if not I will also give you access to the branch of my fork that has this code.

**Comment by lucbv** [comment]
All right the results from the auto-tester are not bad at least.

File: packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2aAlgorithm_kokkos_def.hpp [hunk 1]
```diff
@@ -60,100 +60,253 @@
...
-    std::vector<unsigned> aggStat;
-    aggStat.resize(aggstatHost.extent(0));
-    for(size_t idx = 0; idx < aggstatHost.extent(0); ++idx) {
-      aggStat[idx] = aggstatHost(idx);
+
+    if(params.get<bool>("aggregation: deterministic")) {
+      Monitor m(*this, "BuildAggregatesDeterministic");
+      BuildAggregatesDeterministic(params, graph, aggregates, aggStat, numNonAggregatedNodes);
+    } else {
+      Monitor m(*this, "BuildAggregates");
```
Comment (jhux2) [inline packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2aAlgorithm_kokkos_def.hpp]: Hah, don't want to worry users :).

**Comment by brian-kelley** [comment]
@jhux2 Yes, all the phases are tested in JustDist2DeterUncoupledAggregation in unit_tests_kokkos/Aggregates_kokkos.cpp.

**Comment by lucbv** [comment]
Removing the WIP flag to have the auto-tester try this again.
I am expecting things to work fine in which case this will be ready to be merged.

File: packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2bAlgorithm_kokkos_def.hpp [hunk 2]
```diff
@@ -66,104 +66,241 @@ namespace MueLu {
...
     // Assuming that the aggregate size is 3 in each direction running the algorithm only twice
     // should be sufficient.
-    for (int k = 0; k < 2; k++) {
-      for (LO i = 0; i < numRows; i++) {
-        if (aggStat[i] != READY)
-          continue;
+    // lbv: If the prior phase of aggregation where run without specifying an aggregate size,
+    // the distance 2 coloring and phase 1 aggregation actually guarantee that only one iteration
+    // is needed to reach distance 2 neighbors.
+    int maxIters = 2;
```
Comment (jhux2) [inline packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2bAlgorithm_kokkos_def.hpp]: Would it be of interest to allow maxIters>2?  I'm not advocating adding an option now, I'm just thinking to the future.

File: packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2bAlgorithm_kokkos_def.hpp [hunk 3]
```diff
@@ -66,104 +66,241 @@ namespace MueLu {
...
+
+    Kokkos::deep_copy(connectWeight, defaultConnectWeight);
+
+    // We do this cycle twice.
+    // I don't know why, but ML does it too
+    // taw: by running the aggregation routine more than once there is a chance that also
+    // non-aggregated nodes with a node distance of two are added to existing aggregates.
+    // Assuming that the aggregate size is 3 in each direction running the algorithm only twice
+    // should be sufficient.
+    int maxIters = 2;
```
Comment (jhux2) [inline packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2bAlgorithm_kokkos_def.hpp]: Same comment as above.

**Pull request merged**